### ISSUE TITLE
Savestates save or load with threaded tasks off slows down in proportion to the console savestate size.

### ISSUE THREAD

**Issue by i30817** [libretro/RetroArch#14377]
## Description

The title means that savestates and loadstates get _slower_ the more memory the savestate has nonlinearly. While a 'small' savestate, from the nes for instance, might take two seconds, the ps1 takes _50_ seconds, the ps2 several minutes etc. 

This can easily be confirmed by turning off threaded tasks in the advanced settings -> UI menu then attempting to save in any ps1 game:

[Peek 2022-09-03 20-21.webm](https://user-images.githubusercontent.com/757900/188285149-66cbc6fc-ce0b-48e2-ab9c-6cadd916fc9d.webm)

It can also be also checked by installing bpftrace `sudo apt install bpftrace` then using it to record the number of times the savestate dispatch functions are called and the interval of time between them. This delay _also_ occurs during load state so these interruptions severely disrupt IO on both cases.

(these retroarch paths are where i keep my master build)

`sudo bpftrace -e 'uprobe:/home/i3/Documents/Projects/RetroArch/retroarch:task_push_save_state { @start=nsecs;} uprobe:/usr/bin/retroarch:task_save_handler { @probe = count(); } uprobe:/usr/bin/retroarch:task_save_handler_finished { printf("elapsed ms %d:\n", (nsecs -@start) / 1000000); exit(); } '`

### Expected behavior

I expected that the number of calls that `task_save_handler` is fixed (so  SAVE_STATE_CHUNK is _not_ hardcoded, but variable based on the platform savestate size) based, or this whole 'segment the savestate to display a slider` thing _does not happen_ when threaded tasks is turned of, just a final 'saved state' message, or a persistent one that is replaced by 'saved' at the end, so two calls.

### Actual behavior

The nonsense in the video occurs, in the platforms that can afford it the least; and if you'd like retroarch not to have a extra UI thread used for basically nothing, you suffer this on the savestates.

### Steps to reproduce the bug

1. Turn off threaded tasks
2. Load a game of any platform at 16+ bits
3. save or load a savestate

### Bisect Results

This always happened

### Version/Commit
You can find this information under Information/System Information

- RetroArch: `RetroArch: Frontend for libretro -- v1.10.3 -- 3abd414656 --`

### Environment information

- OS: Ubuntu 22.04.1
- Compiler: `GCC (11.2.0) 64-bit Built: Sep  3 2022`

This is a duplicate of a bug that was close inexplicably since this problem is _very much_ not fixed or irrelevant.

**Comment by i30817** [comment]
This is actually slightly surprising in that i expected if the threaded tasks were turned off, that a savestate would show a lot of delays in the emulation. 

This doesn't appear to be a thing that happens, which makes me think that the savestate already is being written in a another thread than the core main thread. This is fine, what's not fine is that writing being delayed because of a accident of savestate size.

**Comment by LibretroAdmin** [comment]
So I might have to refresh my memory but I recall when we did the task queue implementation, my thought was that if the task queue was not available on a platform because of no threading, that these non-threaded 'tasks' should not 'block' or interrupt the UI or even slow down the gameplay much. Hence why some kinda dodgy heuristic was chosen to split up the workload in 'chunks' based on what would fit into a frame without ruining the framerate too much.

There might be better ways of going about this and it might be that the heuristic that was used for this was total amateur level and we might have to do something based on disk I/O performance and some other heuristics, but at least the thought behind it is/was decent. If you have any suggestions on how to improve it while still abiding by this design ideal, I'm all ears.

I believe even image decoding works on a similar principle although I believe there we need some improvements made, PNG decoding tends to take way too long right now in a weird way (since it's not bottlenecked by disk I/O or even CPU as far as I've been able to tell).

TLDR - we need more intelligent ways of deciding chunksizes for non-threaded tasks.

**Comment by i30817** [comment]
Worst case possible.

Well, let's just say i don't agree with those limitations. If i was in a platform that had no threads i wouldn't want to my loadstates/savestates to literally take minutes at a time (and this is in a _recent_ computer).

So I'd abandon that and pause emulation during savestate/saveload (not for downloads though), and use larger SAVE_STATE_CHUNK always (so that it gives a fixed number of iterations, say one every 0.5% of the total).

That way, saves 'still are fast', they're slightly faster for threaded computers (because they write more), there are 200 iterations always and the people that need the speed the most don't get to wait anxiously for the savestate/saveload to load forever for a 16 mb savestate.

That's my opinion.

**Comment by i30817** [comment]
This bug still bothers me but I've since accepted it exists because RetroArch doesn't actually have a thread apart for making savestates when treaded tasks are disabled, so instead of pausing emulation it does this. 

I don't accept the rationale though, or its necessity because of the absurdly bad way that segmenting I\O writes degenerates performance. Even when I had that old computer (I have a even worse now lol) if I compared dolphin upstream and dolphin core savestates WITH threaded tasks for the core... well the core was slower. By seconds. Dolphin was like two seconds to save, and this was on a rotational hard disk. Dolphin core was more like... 5.

Modern I\O is speculative. If you start writing a array to disk every part of the system warms up in the expectation that you will indeed write the whole thing out, and context switches are not going to happen and the data will sit in various minimal buffers waiting to be beamed out to the SSD or HDD. 

Unless you interrupt it. This is why dolphin oh so immersion breaking pause  is a better experience than RA segmented write (well besides the absurdity of the fixed value chosen for the fps pause apparently - a rationale I find suspect in itself, because of that warmup a cpu writing to disk raw array data is almost certainly going to pipeline that stuff in a way it can do something else). You could press the button and go right to gameplay and press a button again without fear that you're still writing out a savestate and you new one would happen much later than expected.
